test/arrayops.jl

```diff
@@ -2524,3 +2524,29 @@ Base.view(::T25958, args...) = args
     @test t[end,end,1]   == @view(t[end,end,1])   == @views t[end,end,1]
     @test t[end,end,end] == @view(t[end,end,end]) == @views t[end,end,end]
 end
+
+@testset "collect on iterator with incorrect length" begin
+    # Iterator with declared length too large
+    struct InvalidIter1 end
+    Base.length(::InvalidIter1) = 2
+    Base.iterate(::InvalidIter1, i=1) = i > 1 ? nothing : (i, (i + 1))
+
+    @test_throws ErrorException collect(InvalidIter1())
+    @test_throws ErrorException collect(Any, InvalidIter1())
+    @test_throws ErrorException collect(Int, InvalidIter1())
+    @test_throws ErrorException [x for x in InvalidIter1()]
+    # Should also throw ErrorException
+    @test_broken length(Int[x for x in InvalidIter1()]) != 2
+
+    # Iterator with declared length too small
+    struct InvalidIter2 end
+    Base.length(::InvalidIter2) = 2
+    Base.iterate(::InvalidIter2, i=1) = i > 3 ? nothing : (i, (i + 1))
+
+    @test_throws ErrorException collect(InvalidIter2())
+    @test_throws ErrorException collect(Any, InvalidIter2())
+    @test_throws ErrorException collect(Int, InvalidIter2())
+    @test_throws ErrorException [x for x in InvalidIter2()]
```

Comment:
I don't think we can test this, since it will do an invalid memory write.

Reply:
Mmm, yeah, so the only way to catch this would be to drop the `@inbounds` in `collect_to!`, or put a `i == len` check as in `copyto!`/`copyto_check_length!`. Not sure that's worth it, but it sounds inconsistent to have it in one place and not the other.

Reply:
Better to add checks for some cases (when it can be efficient) than to have no checks at all, I'd say.

Reply:
But are there reason to care less about performance in `copyto!` than in `collect_to!`?

Reply:
`collect_to!` is used by comprehensions.

Reply:
Hmm, does that mean it's OK for `copyto!` to be slow? :-/

Reply:
No, it doesn't mean it's OK for `copyto!` to be slow, it just means we don't want regressions.

Reply:
That was actually an honest question: maybe we should remove the check from `copyto!` to make it faster.